### PR TITLE
refactor: pass MessageBox params as a struct

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

This work is intended as a precursor to the future implementation of https://github.com/electron/electron/issues/12437.

Historically, we've been passing in all MessageBox parameters individually, which makes augmenting or improving MessageBox functionality challenging because to change or add even one argument requires a huge cascade of argument changes that leaves room for errors.

For other file dialog related APIs, we use a struct (`DialogSettings`), and so this PR takes a similar approach and refactors MessageBox parameters into a struct (`MessageBoxSettings`) which we then use to simplify argument passing and which will enable us to more quickly iterate and improve upon functionality in the future.

cc @nornagon @MarshallOfSound @zcbenz 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

File: atom/browser/ui/message_box_gtk.cc [hunk 1]
```diff
@@ -200,46 +197,26 @@ void GtkMessageBox::OnCheckboxToggled(GtkWidget* widget) {
...
 
 void ShowErrorBox(const base::string16& title, const base::string16& content) {
   if (Browser::Get()->is_ready()) {
-    GtkMessageBox(nullptr, MessageBoxType::kError, {"OK"}, -1, 0, "Error",
-                  base::UTF16ToUTF8(title).c_str(),
-                  base::UTF16ToUTF8(content).c_str(), "", false,
-                  gfx::ImageSkia())
-        .RunSynchronous();
+    atom::MessageBoxSettings settings;
+    settings.type = 3;  // MessageBoxType::kError
```
Comment (ckerr) [inline atom/browser/ui/message_box_gtk.cc]: Why not `settings.type = MessageboxType::kError`?

If that doesn't work and `type` _has_ to be an int, a backup plan would be to cast MessageBoxType::kError to an int here

Either way we should not be hardcoding the number

File: atom/browser/ui/message_box_gtk.cc [hunk 2]
```diff
@@ -200,46 +197,26 @@ void GtkMessageBox::OnCheckboxToggled(GtkWidget* widget) {
...
+    settings.buttons = {"OK"};
+    settings.title = "Error";
+    settings.message = base::UTF16ToUTF8(title).c_str();
+    settings.detail = base::UTF16ToUTF8(content).c_str();
```
Comment (ckerr) [inline atom/browser/ui/message_box_gtk.cc]: `base::UTF16ToUTF8(foo)` returns a std::string, so we should be able to take it directly, e.g.

```
settings.message = base::UTF16ToUTF8(title);
settings.detail = base::UTF16ToUTF8(content);
```

File: atom/browser/ui/message_box_gtk.cc [hunk 3]
```diff
@@ -200,46 +197,26 @@ void GtkMessageBox::OnCheckboxToggled(GtkWidget* widget) {
...
+    settings.icon = gfx::ImageSkia();
```
Comment (ckerr) [inline atom/browser/ui/message_box_gtk.cc]: `settings.icon` is already an empty `gfx::ImageSkia` so this line can be removed.

File: atom/browser/ui/message_box_win.cc [hunk 4]
```diff
@@ -183,98 +187,57 @@ int ShowTaskDialogUTF16(NativeWindow* parent,
...
   std::vector<base::string16> utf16_buttons;
-  for (const auto& button : buttons)
+  for (const auto& button : settings.buttons)
     utf16_buttons.push_back(base::UTF8ToUTF16(button));
 
+  base::string16 title_16 = base::UTF8ToUTF16(settings.title);
+  base::string16 message_16 = base::UTF8ToUTF16(settings.message);
+  base::string16 detail_16 = base::UTF8ToUTF16(settings.detail);
+  base::string16 checkbox_label_16 = base::UTF8ToUTF16(settings.checkbox_label);
+  bool checkbox_checked16 = settings.checkbox_checked;
```
Comment (ckerr) [inline atom/browser/ui/message_box_win.cc]: _(nit)_ all the above can be `const`

_(nit)_ the variable name "checkbox_checked16" is fun. Converting a bool to a 16-bit bool is an interesting choice :neutral_face:

File: atom/common/native_mate_converters/message_box_converter.cc [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  dict.Get("buttons", &(out->buttons));
+  dict.Get("defaultId", &(out->default_id));
+  dict.Get("cancelId", &(out->cancel_id));
+  dict.Get("options", &(out->options));
+  dict.Get("title", &(out->title));
+  dict.Get("message", &(out->message));
+  dict.Get("detail", &(out->detail));
+  dict.Get("checkboxLabel", &(out->checkbox_label));
+  dict.Get("checkboxChecked", &(out->checkbox_checked));
+  dict.Get("icon", &(out->icon));
```
Comment (ckerr) [inline atom/common/native_mate_converters/message_box_converter.cc]: _nit_: The parens are unnecessary here. `&out->icon` is fine

File: atom/browser/ui/message_box.h [hunk 6]
```diff
@@ -32,32 +29,31 @@ enum MessageBoxOptions {
...
+  atom::NativeWindow* parent_window = nullptr;
+  int type;
+  std::vector<std::string> buttons;
+  int default_id;
+  int cancel_id;
+  int options;
+  std::string title;
+  std::string message;
+  std::string detail;
+  std::string checkbox_label = "";
```
[the remark this answers is not in the document]
Comment (codebytere) [reply atom/browser/ui/message_box.h]: (resolved IRL, not needed!)

File: atom/browser/ui/message_box.h [hunk 7]
```diff
@@ -32,32 +29,31 @@ enum MessageBoxOptions {
...
-                       int options,
-                       const std::string& title,
-                       const std::string& message,
-                       const std::string& detail,
-                       const gfx::ImageSkia& icon);
+struct MessageBoxSettings {
+  atom::NativeWindow* parent_window = nullptr;
+  int type;
+  std::vector<std::string> buttons;
+  int default_id;
```
[the remark this answers is not in the document]
Comment (codebytere) [reply atom/browser/ui/message_box.h]: We handle this in custom fashion on macOS at least, so i think we may want to leave this as-is 🤔 @MarshallOfSound you may know more?

[the remark this answers is not in the document]
Comment (MarshallOfSound) [reply atom/browser/ui/message_box.h]: A particular default chosen will be chosen in JS land, this should always receive a value and shouldn't need a default.  However the converter should throw / fail if this value is not provided.

File: atom/common/native_mate_converters/message_box_converter.cc [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+#include "atom/common/native_mate_converters/message_box_converter.h"
+
+#include "atom/browser/api/atom_api_browser_window.h"
+#include "atom/common/native_mate_converters/file_path_converter.h"
+#include "atom/common/native_mate_converters/image_converter.h"
+#include "native_mate/dictionary.h"
+
+namespace mate {
+
+bool Converter<atom::MessageBoxSettings>::FromV8(
```
Comment (MarshallOfSound) [inline atom/common/native_mate_converters/message_box_converter.cc]: Some of the values here are 100% required and should never be undefined or empty.  The converter should fail for instance if `defaultId` and `cancelId` are not set.  Probably things like `title` as well.

File: atom/browser/ui/message_box.h [hunk 9]
```diff
@@ -32,32 +29,31 @@ enum MessageBoxOptions {
...
-                       int default_id,
-                       int cancel_id,
-                       int options,
-                       const std::string& title,
-                       const std::string& message,
-                       const std::string& detail,
-                       const gfx::ImageSkia& icon);
+struct MessageBoxSettings {
+  atom::NativeWindow* parent_window = nullptr;
+  int type;
```
Comment (miniak) [inline atom/browser/ui/message_box.h]: why `int`?

File: atom/browser/ui/message_box_gtk.cc (hunk 1 above)
Comment (miniak) [reply atom/browser/ui/message_box_gtk.cc]: shouldn't `type` be `MessageboxType`?

Comment (codebytere) [reply atom/browser/ui/message_box_gtk.cc]: done

File: atom/browser/ui/message_box.h [hunk 10]
```diff
@@ -32,32 +29,31 @@ enum MessageBoxOptions {
...
-                       int default_id,
-                       int cancel_id,
-                       int options,
-                       const std::string& title,
-                       const std::string& message,
-                       const std::string& detail,
-                       const gfx::ImageSkia& icon);
+struct MessageBoxSettings {
+  atom::NativeWindow* parent_window = nullptr;
+  MessageBoxType type;
```
Comment (miniak) [inline atom/browser/ui/message_box.h]: `= atom::MessageBoxType::kNone`?

File: atom/browser/atom_javascript_dialog_manager.cc [hunk 11]
```diff
@@ -91,10 +91,18 @@ void AtomJavaScriptDialogManager::RunJavaScriptDialog(
       window = relay->GetNativeWindow();
   }
 
+  atom::MessageBoxSettings settings;
+  settings.parent_window = window;
+  settings.type = atom::MessageBoxType::kNone;
+  settings.buttons = buttons;
+  settings.default_id = default_id;
+  settings.cancel_id = cancel_id;
+  settings.icon = gfx::ImageSkia();
```
Comment (miniak) [inline atom/browser/atom_javascript_dialog_manager.cc]: this is not necessary

File: atom/browser/atom_javascript_dialog_manager.cc [hunk 12]
```diff
@@ -91,10 +91,18 @@ void AtomJavaScriptDialogManager::RunJavaScriptDialog(
       window = relay->GetNativeWindow();
   }
 
+  atom::MessageBoxSettings settings;
+  settings.parent_window = window;
+  settings.type = atom::MessageBoxType::kNone;
```
Comment (miniak) [inline atom/browser/atom_javascript_dialog_manager.cc]: I would use `atom::MessageBoxType::kNone` as default value in the struct instead

File: atom/browser/ui/message_box.h [hunk 13]
```diff
@@ -32,32 +29,31 @@ enum MessageBoxOptions {
...
+  std::vector<std::string> buttons;
+  int default_id;
+  int cancel_id;
+  int options = 0;
```
Comment (miniak) [inline atom/browser/ui/message_box.h]: = `atom::MessageBoxOptions::MESSAGE_BOX_NONE`?

File: atom/browser/atom_javascript_dialog_manager.cc [hunk 14]
```diff
@@ -91,10 +91,18 @@ void AtomJavaScriptDialogManager::RunJavaScriptDialog(
...
+  settings.buttons = buttons;
+  settings.default_id = default_id;
+  settings.cancel_id = cancel_id;
+  settings.icon = gfx::ImageSkia();
+  settings.options = atom::MessageBoxOptions::MESSAGE_BOX_NONE;
```
Comment (miniak) [inline atom/browser/atom_javascript_dialog_manager.cc]: this is not necessary, we already set it to 0 by default

**Pull request merged**